### PR TITLE
Redesign the exposed cache api.

### PR DESCRIPTION
**Pull request by lachlansneff**
This exposes a `Cache` trait and a `FileSystemCache` type that implements that cache trait.

TODO:

- [x] Documentation needs to be written.

### CODE REVIEW

File: lib/runtime/src/lib.rs [hunk 1]
```diff
@@ -154,38 +147,35 @@ pub fn compile(wasm: &[u8]) -> error::CompileResult<Module> {
...
+// /// let cache = compile_cache(wasm)?;
+// /// # Ok(())
+// /// # }
+// /// ```
+// pub fn compile_cache(wasm: &[u8]) -> error::CompileResult<Cache> {
+//     let default_compiler = default_compiler();
+
+//     wasmer_runtime_core::compile_to_cache_with(wasm, default_compiler)
+//         .map(|core_cache| Cache(core_cache))
+// }
```
[the remark this answers is not in the document]
Comment (lachlansneff) [reply lib/runtime/src/lib.rs]: Whoops, will do.

File: lib/runtime-core/Cargo.toml [hunk 2]
```diff
@@ -17,26 +16,25 @@ lazy_static = "1.2.0"
 indexmap = "1.0.2"
 errno = "0.2.4"
 libc = "0.2.48"
+hex = "0.3.2"
 
 # Dependencies for caching.
 [dependencies.serde]
 version = "1.0"
-optional = true
+features = ["rc"]
```
[the remark this answers is not in the document]
Comment (lachlansneff) [reply lib/runtime-core/Cargo.toml]: rc stands for "reference-counted". Serde requires a special feature to be able to serialize/deserialize `Rc` and `Arc`.

[the remark this answers is not in the document]
Comment (syrusakbary) [reply lib/runtime-core/Cargo.toml]: Can we add that as a comment on the `Cargo.toml` ?

[the remark this answers is not in the document]
Comment (lachlansneff) [reply lib/runtime-core/Cargo.toml]: @syrusakbary Okay.

**Pull request merged**